components/butterfly/src/rumor.rs

```diff
@@ -333,6 +388,46 @@ mod storage {
             }
             result
         }
+
+        pub fn keys_rsr(&self) -> Vec<RumorKey> {
+            self.list
+                .read()
+                .values()
+                .flat_map(HashMap::values)
+                .map(RumorKey::from)
+                .collect()
+        }
+    }
+
+    impl<T: Transient> RumorStore<T> {
+        pub fn insert_transient_rsw(&self, rumor: T) -> bool {
+            if !rumor.expired(Utc::now()) {
+                self.insert_rsw(rumor)
+            } else {
+                false
+            }
+        }
+
+        /// Remove all rumors that have expired from our rumor store.
+        pub fn purge_expired_rsr(&self, expiration_date: DateTime<Utc>) {
```

Comment:
I'm not sure how this works. It's trying to take a write lock while still holding a read lock.

Rather than take this approach (which also needs a separate lookup for each key), I think it makes sense to add a `RumorStore::lock_rsw`, which will return an iterable guard that we can use for mutation. Then implement something like
```rust
    impl<'a, T: Transient> IterableWriteGuard<'a, RumorMap<T>> {
        pub fn purge_expired_by(&mut self, time: DateTime<Utc>) {
            for submap in self.0.values_mut() {
                submap.retain(|_, r| !r.expired_by(time))
            }
        }
    }
```
and the callers would look like:
```rust
rumor_store.lock_rsw().purge_expired_by(time);
```

Reply:
Will do